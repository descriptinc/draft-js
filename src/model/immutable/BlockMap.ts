/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import {BlockNodeRecord} from './BlockNodeRecord';
import {flatMap, flatten} from '../descript/Iterables';

export type BlockMap = ReadonlyMap<string, BlockNodeRecord>;

// FIXME [mvp]: dealing with order of `updates`

export function mergeMapUpdates<T>(
  originalMap: ReadonlyMap<string, T>,
  updates: Record<string, T | null>,
): ReadonlyMap<string, T> {
  let didChange = false;
  const seenKeys = new Set<string>();
  const result: ReadonlyMap<string, T> = new Map(
    flatten([
      flatMap(originalMap, (entry): [string, T] | undefined => {
        const [key, existingBlock] = entry;
        seenKeys.add(key);
        const newBlock = updates[key];
        if (newBlock === null) {
          // remove this block from the map
          return undefined;
        }
        if (!newBlock || newBlock === existingBlock) {
          return entry;
        }
        didChange = true;
        return [key, newBlock];
      }),
      flatMap(Object.entries(updates), (entry): [string, T] | undefined => {
        const [key, block] = entry;
        if (!block || seenKeys.has(key)) {
          return undefined;
        }
        didChange = true;
        return [key, block];
      }),
    ]),
  );
  return didChange ? result : originalMap;
}
