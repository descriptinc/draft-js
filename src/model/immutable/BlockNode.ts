/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {DraftBlockType} from '../constants/DraftBlockType';
import {CharacterMetadata} from './CharacterMetadata';

export type BlockNodeKey = string;

export type BlockNode = Readonly<{
  characterList: readonly CharacterMetadata[];
  data: Readonly<Record<string, any>>;
  depth: number;
  key: BlockNodeKey;
  text: string;
  type: DraftBlockType;
}>;
