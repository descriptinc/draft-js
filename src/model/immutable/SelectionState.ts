/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

export type SelectionState = Readonly<{
  anchorKey: string;
  anchorOffset: number;
  focusKey: string;
  focusOffset: number;
  isBackward: boolean;
  hasFocus: boolean;
}>;

export function getStartOffset({
  isBackward,
  focusOffset,
  anchorOffset,
}: SelectionState): number {
  return isBackward ? focusOffset : anchorOffset;
}

export function getEndOffset({
  isBackward,
  anchorOffset,
  focusOffset,
}: SelectionState): number {
  return isBackward ? anchorOffset : focusOffset;
}

export function isCollapsed({
  anchorKey,
  focusKey,
  anchorOffset,
  focusOffset,
}: SelectionState): boolean {
  return anchorKey === focusKey && anchorOffset === focusOffset;
}

export function getStartKey({
  isBackward,
  focusKey,
  anchorKey,
}: SelectionState): string {
  return isBackward ? focusKey : anchorKey;
}

export function getEndKey({
  isBackward,
  focusKey,
  anchorKey,
}: SelectionState): string {
  return isBackward ? anchorKey : focusKey;
}

/**
 * Return whether the specified range overlaps with an edge of the
 * SelectionState.
 */
export function hasEdgeWithin(
  sel: SelectionState,
  blockKey: string,
  start: number,
  end: number,
): boolean {
  const {anchorKey, focusKey, anchorOffset, focusOffset} = sel;
  if (anchorKey === focusKey && anchorKey === blockKey) {
    const selectionStart = getStartOffset(sel);
    const selectionEnd = getEndOffset(sel);
    return (
      (start <= selectionStart && selectionStart <= end) || // selectionStart is between start and end, or
      (start <= selectionEnd && selectionEnd <= end) // selectionEnd is between start and end
    );
  }

  if (blockKey !== anchorKey && blockKey !== focusKey) {
    return false;
  }

  const offsetToCheck = blockKey === anchorKey ? anchorOffset : focusOffset;

  return start <= offsetToCheck && end >= offsetToCheck;
}

export function makeSelectionState({
  anchorKey,
  anchorOffset,
  focusKey,
  focusOffset,
  isBackward = false,
  hasFocus = false,
}: Partial<SelectionState> &
  Pick<
    SelectionState,
    'anchorKey' | 'anchorOffset' | 'focusKey' | 'focusOffset'
  >): SelectionState {
  return {
    anchorKey,
    anchorOffset,
    focusKey,
    focusOffset,
    isBackward,
    hasFocus,
  };
}

export function makeNullSelection(): SelectionState {
  return makeEmptySelection('');
}

export function makeEmptySelection(key: string): SelectionState {
  return makeSelectionState({
    anchorKey: key,
    anchorOffset: 0,
    focusKey: key,
    focusOffset: 0,
  });
}
