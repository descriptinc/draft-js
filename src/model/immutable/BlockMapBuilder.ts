/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+draft_js
 */

import {BlockMap} from './BlockMap';
import {map} from '../descript/Iterables';
import {BlockNode} from './BlockNode';

export function createFromArray(blocks: Iterable<BlockNode>): BlockMap {
  return new Map(
    map(blocks, (block): [string, BlockNode] => [block.key, block]),
  );
}
