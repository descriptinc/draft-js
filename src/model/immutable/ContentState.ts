/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {BlockMap} from './BlockMap';
import {
  SelectionState,
  makeEmptySelection,
  makeNullSelection,
} from './SelectionState';
import {
  first,
  join,
  last,
  map,
  rest,
  skipUntil,
  takeUntil,
} from '../descript/Iterables';
import {DraftEntityType} from '../entity/DraftEntityType';
import {DraftEntityMutability} from '../entity/DraftEntityMutability';
import {createFromArray} from './BlockMapBuilder';
import {makeContentBlock} from './ContentBlock';
import DraftEntity, {DraftEntityMapObject} from '../entity/DraftEntity';
import {DraftEntityInstance} from '../entity/DraftEntityInstance';
import sanitizeDraftText from '../encoding/sanitizeDraftText';
import {BlockNode} from './BlockNode';

export type ContentState = Readonly<{
  blockMap: BlockMap;
  selectionBefore: SelectionState;
  selectionAfter: SelectionState;
}>;

export function makeContentState({
  blockMap,
  selectionAfter = makeEmptySelection(first(blockMap.keys())!),
  selectionBefore = makeEmptySelection(first(blockMap.keys())!),
}: Partial<ContentState> & Pick<ContentState, 'blockMap'>): ContentState {
  // FIXME [perf]: force type to be exact and don't create a new object?
  return {
    blockMap,
    selectionBefore,
    selectionAfter,
  };
}

export function getLastCreatedEntityKey(): string {
  // TODO: update this when we fully remove DraftEntity
  return DraftEntity.__getLastCreatedEntityKey();
}

export function getBlockForKey(
  {blockMap}: ContentState,
  key: string,
): BlockNode {
  const block = blockMap.get(key);
  if (!block) {
    throw new Error('No block found for key');
  }
  return block;
}

export function getFirstBlock({blockMap}: ContentState): BlockNode {
  const block = first(blockMap.values());
  if (!block) {
    throw new Error('Block map is empty');
  }
  return block;
}

export function getLastBlock({blockMap}: ContentState): BlockNode {
  const block = last(blockMap.values()); // FIXME [perf]: O(n)
  if (!block) {
    throw new Error('Block map is empty');
  }
  return block;
}

export function hasText(content: ContentState): boolean {
  const {blockMap} = content;
  if (blockMap.size > 1) {
    return true;
  }
  return (
    blockMap.size > 1 ||
    (blockMap.size > 0 &&
      escape(getFirstBlock(content).text).replace(/%u200B/g, '').length > 0)
  );
}

export function getPlainText(content: ContentState, separator = '\n'): string {
  return join(
    map(content.blockMap.values(), b => b.text),
    separator,
  );
}

export function createEntity(
  type: DraftEntityType,
  mutability: DraftEntityMutability,
  data?: Record<string, unknown> | null,
): string {
  // TODO: update this when we fully remove DraftEntity
  return DraftEntity.__create(type, mutability, data);
}

export function mergeEntityData(
  key: string,
  toMerge: Record<string, any>,
): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__mergeData(key, toMerge);
}

export function replaceEntityData(
  key: string,
  newData: Record<string, any>,
): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__replaceData(key, newData);
}

export function addEntity(instance: DraftEntityInstance): void {
  // TODO: update this when we fully remove DraftEntity
  DraftEntity.__add(instance);
}

export function getEntity(key: string): DraftEntityInstance {
  // TODO: update this when we fully remove DraftEntity
  return DraftEntity.__get(key);
}

export function createFromBlockArray(
  blocks: BlockNode[] | {contentBlocks: BlockNode[]},
): ContentState {
  // TODO: remove this when we completely deprecate the old entity API
  const theBlocks = Array.isArray(blocks) ? blocks : blocks.contentBlocks;

  const blockMap = createFromArray(theBlocks);
  const firstBlock = first(blockMap.values());
  const selectionState = !firstBlock
    ? makeNullSelection()
    : makeEmptySelection(firstBlock.key);

  return makeContentState({
    blockMap,
    selectionBefore: selectionState,
    selectionAfter: selectionState,
  });
}

export function createFromText(
  text: string,
  delimiter: string | RegExp = /\r\n?|\n/g,
): ContentState {
  const strings = text.split(delimiter);
  const blocks = strings.map(block => {
    block = sanitizeDraftText(block);

    return makeContentBlock({
      text: block,
    });
  });
  return createFromBlockArray(blocks);
}

export function getBlockBefore(
  {blockMap}: ContentState,
  blockKey: string,
): BlockNode | undefined {
  // FIXME [perf]: cache
  const before = last(takeUntil(blockMap, ([k]) => k === blockKey));
  return before?.[1];
}

export function getBlockAfter(
  {blockMap}: ContentState,
  blockKey: string,
): BlockNode | undefined {
  // FIXME [perf]: cache
  const after = first(rest(skipUntil(blockMap, ([k]) => k === blockKey)));
  return after?.[1];
}

export function getEntityMap(_: ContentState): DraftEntityMapObject {
  return DraftEntity;
}
