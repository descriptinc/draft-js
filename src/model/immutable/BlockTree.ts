/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+draft_js
 */

'use strict';

import {ContentState} from './ContentState';
import {BlockNodeRecord} from './BlockNodeRecord';
import {DraftDecoratorType} from '../decorators/DraftDecoratorType';
import {getRange} from '../descript/RangeCache';
import {findRangesImmutable} from './findRangesImmutable';
import {map} from '../descript/Iterables';
import {CharacterMetadata} from './CharacterMetadata';

const returnTrue = function() {
  return true;
};

type LeafRange = Readonly<{
  start: number;
  end: number;
}>;

type DecoratorRange = LeafRange &
  Readonly<{
    decoratorKey: string | null;
    leaves: readonly LeafRange[];
  }>;

function makeDecoratorRange(
  start: number,
  end: number,
  decoratorKey: string | null,
  leaves: readonly LeafRange[],
): DecoratorRange {
  return {
    start,
    end,
    decoratorKey,
    leaves,
  };
}

const BlockTree = {
  /**
   * Generate a block tree for a given ContentBlock/decorator pair.
   */
  generate: function(
    contentState: ContentState,
    block: BlockNodeRecord,
    decorator: DraftDecoratorType | null,
  ): readonly DecoratorRange[] {
    const textLength = block.text.length;
    if (!textLength) {
      return [makeDecoratorRange(0, 0, null, [getRange(0, 0)])];
    }

    const leafSets: DecoratorRange[] = [];
    const decorations = decorator
      ? decorator.getDecorations(block, contentState)
      : new Array(textLength).fill(null);

    const chars = block.characterList;

    findRangesImmutable(decorations, areEqual, returnTrue, (start, end) => {
      leafSets.push(
        makeDecoratorRange(
          start,
          end,
          decorations[start],
          generateLeaves(chars.slice(start, end), start),
        ),
      );
    });

    return leafSets;
  },
};

/**
 * Generate LeafRange records for a given character list.
 */
function generateLeaves(
  characters: readonly CharacterMetadata[],
  offset: number,
): readonly LeafRange[] {
  const leaves: LeafRange[] = [];
  findRangesImmutable(
    map(characters, ({style}) => style),
    areEqual,
    returnTrue,
    (start, end) => {
      leaves.push(getRange(start + offset, end + offset));
    },
  );
  return leaves;
}

function areEqual(a: any, b: any): boolean {
  return a === b;
}

module.exports = BlockTree;
