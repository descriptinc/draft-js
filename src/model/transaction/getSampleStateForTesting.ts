/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import {
  createWithContent,
  EditorState,
  forceSelection,
} from '../immutable/EditorState';
import {ContentState, makeContentState} from '../immutable/ContentState';
import {makeSelectionState, SelectionState} from '../immutable/SelectionState';
import {makeContentBlock} from '../immutable/ContentBlock';
import {repeat} from '../descript/Iterables';
import {makeCharacterMetadata} from '../immutable/CharacterMetadata';
import SampleInlineStyles from '../immutable/SampleDraftInlineStyle';
import {createFromArray} from '../immutable/BlockMapBuilder';
import DraftEntity from '../entity/DraftEntity';

const {BOLD, ITALIC} = SampleInlineStyles;
const ENTITY_KEY = '2';

const BLOCKS = [
  makeContentBlock({
    key: 'a',
    type: 'unstyled',
    text: 'Alpha',
  }),
  makeContentBlock({
    key: 'b',
    type: 'unordered-list-item',
    text: 'Bravo',
    characterList: Array.from(
      repeat(
        5,
        makeCharacterMetadata({
          style: BOLD,
          entity: ENTITY_KEY,
        }),
      ),
    ),
  }),
  makeContentBlock({
    key: 'c',
    type: 'code-block',
    text: 'Test',
  }),
  makeContentBlock({
    key: 'd',
    type: 'code-block',
    text: '',
    characterList: [],
  }),
  makeContentBlock({
    key: 'e',
    type: 'code-block',
    characterList: [],
  }),
  makeContentBlock({
    key: 'f',
    type: 'blockquote',
    text: 'Charlie',
    characterList: Array.from(
      repeat(
        7,
        makeCharacterMetadata({
          style: ITALIC,
          entity: null,
        }),
      ),
    ),
  }),
];

const selectionState = makeSelectionState({
  anchorKey: 'a',
  anchorOffset: 0,
  focusKey: 'a',
  focusOffset: 0,
  isBackward: false,
  hasFocus: true,
});

const blockMap = createFromArray(BLOCKS);
const contentState = makeContentState({
  blockMap,
  selectionBefore: selectionState,
  selectionAfter: selectionState,
});

DraftEntity.__create('IMAGE', 'IMMUTABLE');

let editorState = createWithContent(contentState);
editorState = forceSelection(editorState, selectionState);

const getSampleStateForTesting = (): {
  editorState: EditorState;
  contentState: ContentState;
  selectionState: SelectionState;
} => {
  return {editorState, contentState, selectionState};
};
export default getSampleStateForTesting;
