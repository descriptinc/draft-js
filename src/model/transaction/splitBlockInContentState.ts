/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */
import {ContentState} from '../immutable/ContentState';
import {isCollapsed, SelectionState} from '../immutable/SelectionState';
import invariant from '../../fbjs/invariant';
import generateRandomKey from '../keys/generateRandomKey';
import modifyBlockForContentState from './modifyBlockForContentState';
import {flatten, rest, skipUntil, takeUntil} from '../descript/Iterables';
import {ContentBlock} from '../immutable/ContentBlock';

const splitBlockInContentState = (
  contentState: ContentState,
  selectionState: SelectionState,
): ContentState => {
  invariant(isCollapsed(selectionState), 'Selection range must be collapsed.');

  const key = selectionState.anchorKey;
  const blockMap = contentState.blockMap;
  const blockToSplit = blockMap.get(key)!;
  const text = blockToSplit.text;

  if (!text) {
    const blockType = blockToSplit.type;
    if (
      blockType === 'unordered-list-item' ||
      blockType === 'ordered-list-item'
    ) {
      return modifyBlockForContentState(contentState, selectionState, block => {
        if (block.type !== 'unstyled' || block.depth !== 0) {
          return {
            ...block,
            type: 'unstyled',
            depth: 0,
          };
        }
        return block;
      });
    }
  }

  const offset = selectionState.anchorOffset;
  const chars = blockToSplit.characterList;
  const keyBelow = generateRandomKey();

  const blockAbove = {
    ...blockToSplit,
    text: text.slice(0, offset),
    characterList: chars.slice(0, offset),
  };
  const blockBelow = {
    ...blockAbove,
    key: keyBelow,
    text: text.slice(offset),
    characterList: chars.slice(offset),
    data: new Map(),
  };
  const blocksBefore = takeUntil(
    blockMap,
    ([, block]) => block === blockToSplit,
  );
  const blocksAfter = rest(
    skipUntil(blockMap, ([, block]) => block === blockToSplit),
  );

  const newBlockMap = new Map(
    flatten<[string, ContentBlock]>([
      blocksBefore,
      [
        [key, blockAbove],
        [keyBelow, blockBelow],
      ],
      blocksAfter,
    ]),
  );

  return {
    ...contentState,
    blockMap: newBlockMap,
    selectionBefore: selectionState,
    selectionAfter: {
      ...selectionState,
      anchorKey: keyBelow,
      anchorOffset: 0,
      focusKey: keyBelow,
      focusOffset: 0,
      isBackward: false,
    },
  };
};

export default splitBlockInContentState;
