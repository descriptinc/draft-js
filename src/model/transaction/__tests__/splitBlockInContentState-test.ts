/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 * @format
 */

import getSampleStateForTesting from '../getSampleStateForTesting';
import {SelectionState} from '../../immutable/SelectionState';
import splitBlockInContentState from '../splitBlockInContentState';
import {blockMapToJsonArray} from '../../../util/blockMapToJson';
import {getFirstBlock} from '../../immutable/ContentState';

jest.mock('../../keys/generateRandomKey');

const {contentState, selectionState} = getSampleStateForTesting();

const assertSplitBlockInContentState = (
  selection: SelectionState,
  content = contentState,
) => {
  expect(
    blockMapToJsonArray(splitBlockInContentState(content, selection).blockMap),
  ).toMatchSnapshot();
};

test('must be restricted to collapsed selections', () => {
  expect(() => {
    const nonCollapsed = {...selectionState, focusOffset: 1};
    return splitBlockInContentState(contentState, nonCollapsed);
  }).toThrow();

  expect(() => {
    return splitBlockInContentState(contentState, selectionState);
  }).not.toThrow();
});

test('must split at the beginning of a block', () => {
  assertSplitBlockInContentState(selectionState);
});

test('must split within a block', () => {
  const SPLIT_OFFSET = 3;

  assertSplitBlockInContentState({
    ...selectionState,
    anchorOffset: SPLIT_OFFSET,
    focusOffset: SPLIT_OFFSET,
  });
});

test('must split at the end of a block', () => {
  const SPLIT_OFFSET = getFirstBlock(contentState).text.length;

  assertSplitBlockInContentState({
    ...selectionState,
    anchorOffset: SPLIT_OFFSET,
    focusOffset: SPLIT_OFFSET,
  });
});
