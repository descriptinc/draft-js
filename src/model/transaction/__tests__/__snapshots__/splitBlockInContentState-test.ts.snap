// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must split at the beginning of a block 1`] = `
Array [
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "a",
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "key0",
    "text": "Alpha",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
]
`;

exports[`must split at the end of a block 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "a",
    "text": "Alpha",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "data": Map {},
    "depth": 0,
    "key": "key0",
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
]
`;

exports[`must split within a block 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "a",
    "text": "Alp",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Map {},
    "depth": 0,
    "key": "key0",
    "text": "ha",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
]
`;
